### ISSUE TITLE
Memory didn't get collected in time with julia v1.9rc1

### ISSUE THREAD

**Issue by chengchingwen** [JuliaLang/julia#49026]
I found a potential issue on 1.9 that sometimes the GC didn't collect the object unless a manual call of `GC.gc()`. The smallest code that I can found and share to reproduce this issue is the test code of [NeuralAttentionlib.jl](https://github.com/chengchingwen/NeuralAttentionlib.jl), specifically [this part of test](https://github.com/chengchingwen/NeuralAttentionlib.jl/blob/master/test/collapseddims.jl). By running the Pkg test, on v1.8, that part didn't use more than 3GB of memory, while on v1.9, it take at least 18GB (I kill the process after this number due to limit amount of memory).

The environment:
```julia
julia> versioninfo()
Julia Version 1.9.0-rc1
Commit 3b2e0d8fbc (2023-03-07 07:51 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 12 × Intel(R) Core(TM) i7-7800X CPU @ 3.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake-avx512)
  Threads: 4 on 12 virtual cores
```

**Comment by gbaraldi** [comment]
How much ram does the whole system have? Just as a question.

**Comment by chengchingwen** [comment]
32GB, but I have about 14GB used by other stuff.

**Comment by KristofferC** [comment]
Did you actually get an OOM error?

**Comment by chengchingwen** [comment]
No, because I made a 32GB swap on my HDD. It would start to use the swap mem.

**Comment by chengchingwen** [comment]
This also cause the [CI on Linux to be terminated](https://github.com/chengchingwen/NeuralAttentionlib.jl/actions/runs/5298744810/jobs/9594405541). By manually adding `GC.gc()`, the issue disappear. I specify the specific part of test is just to show that it's not allocating some large objects but a series of small tests.
